File: Webscraping/Scripts/ggi.r
######################################################################
## Filename: scrape-ggi.r
## Description:
## Author: Helge Liebert
## Created: Do Dez  6 13:51:53 2018
## Last-Updated: Mi Jan  9 14:04:51 2019
######################################################################

## load library
library(rvest)

## initiate session on site
site <- "http://nap.psa.gov.ph/ggi/default.asp"
session <- html_session(site)

## look at nodes
html_nodes(session, "form")
html_nodes(session, "option")

## get the drop-down options
options <- html_nodes(session, "select[name='strMunicipality2'] > option")
municipalities <- data.frame(
    value  = html_attr(options, "value"),
    option = html_text(options))
head(municipalities)
dim(municipalities)

## empty data frame to be filled
data <- data.frame(matrix(nrow = 0, ncol = 5))

## get ggi data for all municipalities in the list
for (op in municipalities$value[1:2]) {

    ## display some information
    print(op)
    print(paste0(which(op == municipalities$value), "/",
                 nrow(municipalities), " ", op))

    ## set option and submit form
    form <- html_form(html_node(session, "#form2"))
    form <- set_values(form, "strMunicipality2" = op)
    newpage <- submit_form(session, form)

    ## get table and process it
    table <- html_table(html_node(newpage, "table"), fill = TRUE)
    table$municipality <- op
    table <- table[3:nrow(table), ]

    ## assert correct dimension
    stopifnot(ncol(table) == 6)
    ## print(table)

    ## append to data
    data <- rbind(data, table)
}

## build header and add to data frame
header <- c("indicator", "value2005", "rank2005", "value2008", "rank2008",
            "municipality ")
names(data) <- header

## look at data
dim(data)
head(data)

File: Webscraping/Scripts/ggi.py
#!/usr/bin/python3

from csv import writer
from pprint import pprint
import mechanicalsoup
# mechanicalsoup imports bs
# from bs4 import BeautifulSoup

# site to scrape
site = "http://nap.psa.gov.ph/ggi/default.asp"

# output file
csvfile = open(r'Data/ggi-py.csv','wb')
output = writer(csvfile)

# open site and get menu options
br = mechanicalsoup.StatefulBrowser()
br.open(site)
page = br.get_current_page()
menu = page.find('select', id='strMunicipality2').find_all('option')
# equivalent
# menu = page.select('#strMunicipality2 > option')
# extract options, only using options, not location text
options = [el['value'] for el in menu]
# locations = [el.text for el in menu]

# write header
header = ["indicator", "value2005", "rank2005", "value2008", "rank2008", "municipality "]
output.writerow(header)


# iterate over options, limit iterations
for pos, item in enumerate(options[:5]):

    # feedback on current item and position
    print(pos, len(options), item)

    # open site, select form, submit option
    br.open(site)
    br.select_form('#form2')
    br['strMunicipality2'] = item
    br.submit_selected()

    # extract table from page
    table = br.get_current_page().select('.dataTable')[0]
    # pprint(table)

    # prep and write to csv linewise, slice away header
    for row in table.find_all('tr')[2:]:
        line = []
        for cell in row.find_all('td'):
            line.append(cell.get_text().strip())
        line.append(item)
        assert len(line)==6
        # print(line)
        output.writerow(line)

    # alternatively, this list comprehension extracts the table
    # as a nested list for further processing
    # data = [[cell.text.strip() for cell in row.find_all('td')]
    #             for row in table.find_all('tr')]


# close output file connection
csvfile.close()
